src/main/java/com/cloudbees/jenkins/github/AccessTokenCredential.java

```diff
@@ -0,0 +1,109 @@
+package com.cloudbees.jenkins.github;
+
+import com.cloudbees.plugins.credentials.CredentialsDescriptor;
+import com.cloudbees.plugins.credentials.CredentialsScope;
+import com.cloudbees.plugins.credentials.common.StandardUsernameCredentials;
+import com.cloudbees.plugins.credentials.impl.BaseStandardCredentials;
+import hudson.Extension;
+import hudson.Util;
+import hudson.model.AutoCompletionCandidates;
+import hudson.util.FormValidation;
+import hudson.util.Secret;
+import jenkins.model.Jenkins;
+import org.kohsuke.github.GHAuthorization;
+import org.kohsuke.github.GitHub;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
+ */
+public class AccessTokenCredential extends BaseStandardCredentials implements StandardUsernameCredentials {
+
+    private final String githubServer;
+
+    private final String username;
+
+    private final Secret token;
```

Comment:
So you want to provide a different UI flow, but my understanding is that
the resulting token is used just like a password. So you probably want to
implement `StandardUsernamePasswordCredentials` *because* the token is used
to authenticate *as if it were a password*.

Because you are providing an alternative UI flow it makes sense to be a
separate credential type, but the actual stored credentials (to my
understanding, correct me if I am wrong) is a username and a *"password"*
(much like the app specific passwords that GMail lets you generate with 2FA
enabled... The app you give that "password" to does not know it is a token
and not your actual password)

On Sunday, April 12, 2015, Nicolas De loof <notifications@github.com> wrote:

> In src/main/java/com/cloudbees/jenkins/github/AccessTokenCredential.java
> <https://github.com/jenkinsci/github-plugin/pull/45#discussion_r28201391>:
>
> > +import org.kohsuke.stapler.QueryParameter;
> > +
> > +import java.io.IOException;
> > +import java.util.Arrays;
> > +import java.util.List;
> > +
> > +/**
> > + * @author <a href="mailto:nicolas.deloof@gmail.com <javascript:_e(%7B%7D,'cvml','nicolas.deloof@gmail.com');>">Nicolas De Loof</a>
> > + */
> > +public class AccessTokenCredential extends BaseStandardCredentials implements StandardUsernameCredentials {
> > +
> > +    private final String githubServer;
> > +
> > +    private final String username;
> > +
> > +    private final Secret token;
>
> @stphenc my intent here was to offer option to create a token from jenkins
> UI, using a classic username/password authentication to github token
> generation API. The token then for sure can be considered being some
> username/password item, but the related UI is a distinct one.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/github-plugin/pull/45/files#r28201391>.
>


-- 
Sent from my phone